File: test/comparison/SB_NGG.R


## sampling NGG stick breaking
library(tidyverse)
library(copula)
R_q <- function(q, b, k_n,alpha){
  term1 = exp(-((b/k_n)^alpha)*((1 + q^(1/alpha))^alpha - q))
  #term2 = ((q^(1/alpha))/(1+q^(1/alpha)))^(1-alpha)
  term2 = (1+q^(1/alpha))^(alpha-1)
  #term3 = (b/kn)^(theta + alpha*(n-1) -1)
  R_q = term1*term2
  return(R_q)
}


sample_xi_n<- function(alpha, b, theta, k_n,n){
  repeat{
        r_n <- runif(1,0,1)
        xi_n <- rgamma(1,theta/alpha + n,(b/k_n)^(alpha))
        if (r_n <= R_q(xi_n, b, k_n, alpha)){ 
          return( (b/k_n)*((xi_n)^(1/alpha)))
          }
        }
}

#density

density_phi_n<- function(x, alpha, b, theta, k_n,n){
  term1 = exp(-((b/k_n)^alpha)*(1 + x^(1/alpha))^alpha - x)
  term2 = (1 + x^(1/alpha))^(alpha - 1)
  return (term1*term2)
}
density_xi_n<- function(x, alpha, b, theta, k_n,n){
  term1 = exp(-((b/k_n + x)^alpha))
  term2 = (b/k_n + x)^(alpha - 1)
  term3 = x^(theta + (n-1)*alpha)
  return (term1*term2*term3)
}

xi<- vector("numeric", length=10000)
for(i in 1:10000){
  xi[i]<- sample_xi_n(0.75, 1, 1, 0.5,1)
}
hist(xi)


##Test for the sampling of xi_n
x= seq(0,100,length=5000)
hist(xi, breaks=25,probability = TRUE)
#x= seq(0,100,length=1000)
#plot(x,density_phi_n(x,0.5, 1, 1, 0.5,1))
lines(x,density_xi_n(x,0.75,1,1,0.5,1)*2.5)


SB_NGG<- function(alpha, b, theta,N){
u_n <- vector("numeric",length=N-1)
for(i in 1:(N-1)){
  if (i==1) {
  k_n <- 1 
  } else {
    k_n = prod(1- u_n[1:i-1])
  }
  xi_n <- sample_xi_n(alpha, b, theta, k_n,i)
  z_n <- rgamma(1,1- alpha, b/k_n + xi_n)
  x_n <- retstable(alpha, 1, h = b/k_n + xi_n)
  u_n[i] = z_n/(z_n + x_n)
} 
p <- vector("numeric",length=N)
p[1] <- u_n[1]
for(l in 2:(N-1))p[l] <- prod(1 - u_n[1:(l - 1)])*u_n[l]
p[N] <- prod(1 - u_n)   
p
}

n_tr=1000
p_k <- SB_NGG(.75, 1,7.5,n_tr)
plot(1:n_tr,p_k )

Prior_on_K_SB_NGG<- function(alpha,  b, theta, N_s,N_tr, runs=10^4 ){
  array_nc<-c()
  i=1
  err=0
  while (i<=runs){
    weights_NGG<- SB_NGG(alpha, b, theta,N_tr)
    #plot(1:N_tr, weights_NGG)
    c <- sample(1:N_tr,size=N_s, replace=TRUE, prob=weights_NGG )
    n_c<- length(unique(c))
    #n_c
    array_nc[i]<- n_c
    i=i+1
    err = err +weights_NGG[N_tr]
  }
  p_k = tibble(k=as.numeric(names(table(array_nc))), 
               p_k=as.vector(table(array_nc))/sum(table(array_nc)))
  p_zeros= tibble(k=(1:N_s)[!1:N_s%in%p_k$k], 
                  p_k=rep(0,length((1:N_s)[!1:N_s%in%p_k$k])))
  pk_padded= rbind(p_k, p_zeros) %>% arrange(k)
  E_k=sum((1:N_s) *(pk_padded$p_k))
  V_k = sum((((1:N_s) - E_k)^2) *(pk_padded$p_k))
  return(list(pk=pk_padded$p_k, Ek= E_k,Vk=sqrt(V_k), error = err/runs ))
}

p_ngg_10_25 <- Prior_on_K_SB_NGG(0.25, 1,2.5, 100,300, runs=10^3)
plot(1:100, p_ngg_10_25$pk, type="l")

p_ngg <- Prior_on_K_SB_NGG(0.25, 39.0625,1, 100,300, runs=10^3)
plot(1:100, p_ngg$pk, type="l")


p_ngg_1_025 <- Prior_on_K_SB_NGG(0.25, 1,0.25, 100,100, runs=10^3)
plot(1:100, p_ngg_1_025$pk, type="l")


p_ngg_1_075 <- Prior_on_K_SB_NGG(0.75, 1,0.75, 100,500, runs=10^3)
p_ngg
plot(1:100, p_ngg_1_075$pk, type="l")


p_ngg_10_075 <- Prior_on_K_SB_NGG(0.75, 1, 7.5, 100,2000, runs=10^2)
p_ngg
plot(1:100, p_ngg_10_075$pk, type="l")





p_ngg <- Prior_on_K_SB_NGG(0.5, 1, 0.5, 100,250, runs=10^3)
plot(1:100, p_ngg$pk, type="l")


p_ngg <- Prior_on_K_SB_NGG(0.7, 0.1, 0.1, 100,100, runs=10)
plot(1:100, p_ngg$pk, type="l")

File: test/comparison/SB_NGG.jl
## SB Pkn_NGG

include("~/Documents/GitHub/GibbsTypePriors/src/common_functions.jl")
include("~/Documents/GitHub/GibbsTypePriors/src/Cnk.jl")
include("~/Documents/GitHub/GibbsTypePriors/src/Vnk.jl")
include("~/Documents/GitHub/GibbsTypePriors/src/Expect_Kn.jl")
include("~/Documents/GitHub/GibbsTypePriors/src/Pkn.jl")
include("~/Documents/GitHub/GibbsTypePriors/src/MvInv.jl")
include("~/Documents/GitHub/GibbsTypePriors/test/comparison/NGG_FK_sampling.jl")

using DataFrames, DataFramesMeta, RCall
R"library(tidyverse)
library(latex2exp)
"
using Distributions

R"library(tidyverse)
library(copula)"


function R_q(q, b, k_n, alpha)
    t1 = exp(- ((b/k_n)^alpha)*((1 + q^(1/alpha))^alpha - q))
    t2 =  ((q^(1/alpha))/(1+q^(1/alpha)))^(1-alpha)
    return t1*t2
end

function sample_xi_n(alpha, b, theta, k_n, n)
   u =  Uniform(0,1)
   g = Gamma(theta/alpha +n, 1/((b/k_n)^(alpha)))
   r_n = rand(u,1)[1]
   dz_n  = rand(g,1)[1]
   while r_n > R_q(dz_n, b, k_n,alpha)
       u =  Uniform(0,1)
       r_n = rand(u,1)[1]
       dz_n  = rand(g,1)[1]
       if r_n<= dz_n
        # println("true")
       end
         #return (b/k_n)*((dz_n)^(1/alpha))
   end
   return (b/k_n)*((dz_n)^(1/alpha))
end

function density_xi_n(x, alpha, b, theta, k_n, n)
  term1 = exp(-((b/k_n + x)^alpha))
  term2 = (b/k_n + x)^(alpha - 1)
  term3 = x^(theta + (n-1)*alpha)
  return term1*term2*term3
end

## test xi_n density

x_t = collect(range(0,1.2*10, length=100))
xi = Array{Float64}(undef, 10000)
for i in 1:10000
  xi[i] = sample_xi_n(0.75, 1, 0.75, 1,1)
end

## TEST for xi_n
using Plots
histogram(xi,bins=250)
x_t = collect(range(0,1.2*10^2, length=1000))
plot!(x_t,2000*map(x ->density_xi_n(x,0.75, 1.0, 0.75, 1, 1),x_t))

#histogram(sim.^(1/0.25),bins=250, normalize=true)
plot!(x_t,map(x ->density_xi_n(x,0.25, 1.0, 0.25, 1, 1),x_t)/20)


plot!(x_t,2000*map(x ->density_xi_n(x,0.75, 1.0, 0.75, 1, 1),x_t))

function SB_NGG(alpha_, b, theta,N)
  u_n =Array{Float64}(undef, N-1)
  k_n = 1
  for i in (1:(N-1))
    if i == 1
      k_n  = 1
    else
      k_n =prod(vec(ones(1,i-1)).-u_n[1:(i-1)])
    end
      xi_n = sample_xi_n(alpha_, b, theta, k_n,i)
    println(xi_n)
    R"zn = rgamma(1,1- $alpha_, $b/$k_n + $xi_n)"
    #g = Gamma(1- alpha_, 1/(b/k_n + xi_n))
    #z_n  = rand(g,1)[1]
    z_n  = @rget zn
    #h_n  = b/k_n + xi_n
    R" xn <- retstable($alpha_, 1, h =$b/$k_n + $xi_n,method = 'MH')"
    x_n = @rget xn
    u_n[i] = z_n/ (z_n + x_n)
 end

  p = Array{Float64}(undef, N)
  p[1] = u_n[1]
  for l in 2:(N-1)
    p[l] = prod(vec(ones(1,l - 1)).-u_n[1:(l - 1)])*u_n[l]
  end
  p[N]=prod(vec(ones(1,(N-1))).- u_n)
  return p
end

n_tr=10
p_k = SB_NGG(.25, 1,0.25,n_tr)
plot(1:,p_k)



function Pkn_NGG_SB(n, β, σ, M; runs=10^4)
  theta = 1
  b = (σ*theta*β)^(1/σ)
  alpha_ = σ
    array_clust_num = Array{Int64}(undef, runs)
    for i in 1:runs
        weights_NGG_SB = SB_NGG(alpha_,b, theta, M)
        c = wsample(1:M, weights_NGG_SB, n, replace=true)
        n_c =  length(unique(c))
        array_clust_num[i] = n_c
    end
    return proportions(array_clust_num, n)
end



function Pkn_NGG_SB_1(n, β, σ, M; runs=10^4)
  theta = (σ*β)
  b = 1
  alpha_ = σ
    array_clust_num = Array{Int64}(undef, runs)
    for i in 1:runs
        weights_NGG_SB = SB_NGG(alpha_,b, theta, M)
        c = wsample(1:M, weights_NGG_SB, n, replace=true)
        n_c =  length(unique(c))
        array_clust_num[i] = n_c
    end
    return proportions(array_clust_num, n)
end


pks = Pkn_NGG_SB_1(100,1.0,0.25,250; runs=2*100)

R"
pk_sb_1_25=$pks
"

pk_sb_1_75 = Pkn_NGG_SB_1(100,1.0,0.75,250; runs=2*100)
R"
pk_sb_1_75=$pk_sb_1_75
"

pk_sb_10_25 = Pkn_NGG_SB_1(100,10.0,0.25,250; runs=2*100)
R"
pk_sb_10_25=$pk_sb_10_25
"
pk_sb_10_75 = Pkn_NGG_SB_1(100,10.0,0.75,250; runs=2*100)
R"
pk_sb_10_75=$pk_sb_10_25
"

pk_sb_1_25_1000 = Pkn_NGG_SB_1(1000,1.0,0.25,250; runs=2*100)
R"
pk_sb_1_25_1000=$pk_sb_1_25_1000

"
pk_sb_1_75_1000 = Pkn_NGG_SB_1(1000,1.0,0.75,250; runs=2*100)
R"
pk_sb_1_75_1000=$pk_sb_1_75_1000

"
pk_sb_10_25_1000 = Pkn_NGG_SB_1(1000,10.0,0.25,250; runs=2*100)
R"
pk_sb_10_25_1000=$pk_sb_10_25_1000
"
pk_sb_10_75_1000 = Pkn_NGG_SB_1(1000,10.0,0.75,250; runs=2*100)
R"
pk_sb_10_75_1000=$pk_sb_10_75_1000
"


plot(collect(range(1,1000, length=1000)), pk_sb_10_25_1000)
plot!(collect(range(1,1000, length=1000)), pk_sb_1_75_1000)
E_1 =  pk_sb_10_25_1000|> ar -> map(*, ar, 1:1000) |> sum
E_2 =  pk_sb_10_25_1000|> ar -> map(*, ar, 1:1000) |> sum

pk_sb_10_25_1000 = Pkn_NGG_SB_1(1000,10.0,0.25,250; runs=2*100)



#####################################################################


pk_sb_1_25_100_1k = Pkn_NGG_SB(100,1.0,0.25,1000; runs=2*100)

R"
pk_sb_1_25_100_1k=$pk_sb_1_25_100_1k
save(pk_sb_1_25_100_1k,file ='~/Documents/GitHub/GibbsTypePriors/test/comparison/NGG_sb_1_25_100_1k.Rdata')
"

pk_sb_1_75_100_1k = Pkn_NGG_SB_1(100,1.0,0.75,1000; runs=2*100)

R"
pk_sb_1_75_100_1k=$pk_sb_1_75_100_1k
save(pk_sb_1_75_100_1k,file ='~/Documents/GitHub/GibbsTypePriors/test/comparison/NGG_sb_1_75_100_1k_.Rdata')
"

pk_sb_10_25_100_1k = Pkn_NGG_SB_1(100,10.0,0.25,1000; runs=2*100)

R"
pk_sb_10_25_100_1k=$pk_sb_10_25_100_1k
save(pk_sb_10_25_100_1k,file ='~/Documents/GitHub/GibbsTypePriors/test/comparison/NGG_sb_10_25_100_1k.Rdata')
"

pk_sb_10_75_100_1k = Pkn_NGG_SB_1(100,10.0,0.75,1000; runs=2*100)

R"
pk_sb_10_75_100_1k=$pk_sb_10_75_100_1k
save(pk_sb_10_75_100_1k,file ='~/Documents/GitHub/GibbsTypePriors/test/comparison/NGG_sb_10_75_100_1k.Rdata')
"





pk_sb_1_25_1k_1k = Pkn_NGG_SB_1(1000,1.0,0.25,1000; runs=2*100)

R"
pk_sb_1_25_1k_1k=$pk_sb_1_25_1k_1k
save(pk_sb_1_25_1k_1k,file ='~/Documents/GitHub/GibbsTypePriors/test/comparison/NGG_sb_1_25_1k_1k.Rdata')
"


pk_sb_1_75_1k_1k = Pkn_NGG_SB_1(1000,1.0,0.75,1000; runs=2*100)

R"
pk_sb_1_75_1k_1k=$pk_sb_1_75_1k_1k
save(pk_sb_1_75_1k_1k,file ='~/Documents/GitHub/GibbsTypePriors/test/comparison/NGG_sb_1_75_1k_1k.Rdata')
"

pk_sb_10_25_1k_1k = Pkn_NGG_SB_1(1000,10.0,0.25,1000; runs=2*100)

R"
pk_sb_10_25_1k_1k=$pk_sb_10_25_1k_1k
save(pk_sb_10_25_1k_1k,file ='~/Documents/GitHub/GibbsTypePriors/test/comparison/NGG_sb_10_25_1k_1k.Rdata')
"


pk_sb_10_75_1k_1k = Pkn_NGG_SB_1(1000,10.0,0.75,1000; runs=2*100)

R"
pk_sb_10_75_1k_1k=$pk_sb_10_75_1k_1k
save(pk_sb_10_75_1k_1k,file ='~/Documents/GitHub/GibbsTypePriors/test/comparison/NGG_sb_10_75_1k_1k.Rdata')
"
